CRM/Admin/Form/Setting.php

```diff
@@ -120,7 +135,11 @@ public function buildQuickForm() {
           );
         }
         elseif ($add == 'addSelect') {
-          $this->addElement('select', $setting, ts($props['title']), $options['values'], CRM_Utils_Array::value('html_attributes', $props));
+          $element = $this->addElement('select', $setting, ts($props['title']), $options['values'], CRM_Utils_Array::value('html_attributes', $props));
+          if (defined('CIVICRM_ENVIRONMENT')) {
+            $element->freeze();
+            CRM_Core_Session::setStatus(ts('The environment settings have been disabled because it has been overridden in the settings file.'), ts('Environment settings'), 'info');
+          }
```

Comment:
This looks wrong to me. This function is a generic form generator (in a class extended by 14 child classes) and you're making the assumption that all `select` elements are the environment setting.